modules/core/src/main/java/org/apache/ignite/internal/cluster/ClusterGroupAdapter.java

```diff
@@ -741,6 +799,39 @@ protected Object readResolve() throws ObjectStreamException {
         }
     }
 
+    /**
+     * Container for cluster group state.
+     */
+    private static class ClusterGroupState {
+        /** Calculated node ids. */
+        public final Set<UUID> nodeIds;
```

Comment:
I didn't get for what do you hold this variable? As I see you use it only in nodes() method but you can easily use 'Collection<ClusterNode> nodes' instead. So do you really need this variable?